### DISCUSSION TITLE
Disconnect and Exiting

### DISCUSSION THREAD

**Discussion by JBrakefield** [hbldh/bleak#549] [Q&A]
I can't get my script to exit "gracefully" despite much research on the subject. My program pairs to a BLE device and communicates between it and a serial port. I am using the recommended method from the uart example to call the main function like this:

```python
try:
        asyncio.run(ble_pair(address, serialPort, False))
    except asyncio.CancelledError:
        # task is cancelled on disconnect, so we ignore this error
        pass
```

My disconnect handle looks like this:

```python
def handle_disconnect(_: BleakClient):
        print("Device was disconnected, goodbye.")
        # cancelling all tasks effectively ends the program
        serialPort.close()
        for task in asyncio.all_tasks():
            task.cancel()
```

Why do I still get a "KeyboardInterrupt" error when I use ctrl+c? I need to call this script as a module in a GUI that I'm building, and I need multiple instances going at once (connect to more than one BLE device at a time). I am having issues with BLE disconnecting without warning, and any attempt I have made to reconnect (by putting new connection code inside the handle_disconnect) has failed. Is there an example of reconnection after unexpected disconnection? If my disconnect handle cancels all tasks, could this interrupt another BLE connection instance?

**Comment by dlech** [comment]
Don't cancel all tasks unless you actually want the program to exit. 

I don't think you can reconnect without scanning again in Bleak currently, so I don't think it is possible to do this in the disconnect handler.

**Reply by JBrakefield** [reply]
I find that sometimes my device is disconnected without warning. If I can't reconnect from the handler, how would you approach initiating a reconnection? When you say "scanning again" do you mean calling this again?
 ```async with BleakClient(address, disconnected_callback=handle_disconnect) as client:```

 If the disconnect handler cannot somehow initiate this directly, perhaps it could return a signal to my GUI code to try and establish another connection? For testing purposes, it would be great to have a way for the BLE script to try reconnecting if connection is lost.

**Reply by ghost** [reply]
Perhaps you are trying to reconnect before the disconnection event has fully completed? In my experience of using Bleak, if the peripheral initiates the disconnection, you must wait until the disconnection callback is called before attempting to reconnect. This can be achieved using an asyncio [Event](https://docs.python.org/3/library/asyncio-sync.html#asyncio.Event):

```Python
async def run():
    # Let's assume we are connected to the peripheral through BleakClient 'client'

    def disconnected_callback(c: BleakClient):
        disconnection_complete_event.set()

    async def waiter(event: asyncio.Event):
        await event.wait()

    client.set_disconnected_callback(disconnected_callback)
    disconnection_complete_event = asyncio.Event()
    disconnection_complete = asyncio.create_task(waiter(disconnection_complete_event))

    # Let's assume the peripheral disconnects here

    # Since we're going to attempt a reconnection, wait for the disconnection event to fully complete 
    await disconnection_complete

    # Now, it should be safe to reconnect
    await client.connect()
```
